File: backend-legacy-python/app/vpn/xray_api.py
"""Direct xray API client — async wrapper around xray CLI commands.

Uses asyncio.create_subprocess_exec for non-blocking docker exec calls.
Replaces slow per-stat queries with batch operations.
"""

import asyncio
import json
import logging
import os

logger = logging.getLogger(__name__)

# Inbound tags → email suffix and flow
INBOUND_DEFS: list[dict] = [
    {"tag": "VLESS TCP REALITY", "suffix": "xray", "flow": "xtls-rprx-vision"},
    {"tag": "VLESS XHTTP REALITY", "suffix": "xhttp", "flow": ""},
    {"tag": "VLESS gRPC REALITY", "suffix": "grpc", "flow": ""},
    {"tag": "VLESS WS CDN", "suffix": "ws", "flow": ""},
]

EMAIL_SUFFIXES = [d["suffix"] for d in INBOUND_DEFS]


class XrayAPI:
    """Async xray API client via docker exec CLI."""

    def __init__(self, container: str | None = None, api_port: int | None = None):
        self.container = container or os.environ.get("XRAY_CONTAINER", "xray")
        self.api_port = api_port or int(os.environ.get("XRAY_STATS_PORT", "10085"))
        self._server = f"127.0.0.1:{self.api_port}"
        self._docker_available: bool | None = None  # Lazy-checked on first use

    # ── User Management ──

    async def add_user(self, inbound_tag: str, uuid: str, email: str, flow: str = "") -> bool:
        """Add a single user to an inbound via `xray api adi`."""
        user_obj: dict = {"id": uuid, "email": email}
        if flow:
            user_obj["flow"] = flow
        rc, _, stderr = await self._exec_xray(
            "api", "adi", f"--server={self._server}", inbound_tag, json.dumps(user_obj),
        )
        if rc != 0:
            logger.warning("add_user failed [%s] %s: %s", inbound_tag, email, stderr[:200])
        return rc == 0

    async def remove_user(self, inbound_tag: str, email: str) -> bool:
        """Remove a user from an inbound via `xray api rmu`."""
        rc, _, stderr = await self._exec_xray(
            "api", "rmu", f"--server={self._server}", inbound_tag, email,
        )
        if rc != 0:
            logger.debug("remove_user failed [%s] %s: %s", inbound_tag, email, stderr[:200])
        return rc == 0

    async def add_user_to_all_inbounds(self, uuid: str, username: str, short_id: str = "") -> bool:
        """Add user to all known inbounds (TCP, XHTTP, gRPC, WS) concurrently."""
        tasks = [
            self.add_user(d["tag"], uuid, f"{username}@{d['suffix']}", d["flow"])
            for d in INBOUND_DEFS
        ]
        results = await asyncio.gather(*tasks, return_exceptions=True)
        ok = sum(1 for r in results if r is True)
        if ok < len(INBOUND_DEFS):
            logger.warning("add_user_to_all: %s added to %d/%d inbounds", username, ok, len(INBOUND_DEFS))
        return ok > 0

    async def remove_user_from_all_inbounds(self, username: str) -> bool:
        """Remove user from all inbounds concurrently."""
        tasks = [
            self.remove_user(d["tag"], f"{username}@{d['suffix']}")
            for d in INBOUND_DEFS
        ]
        results = await asyncio.gather(*tasks, return_exceptions=True)
        ok = sum(1 for r in results if r is True)
        return ok > 0

    # ── Stats ──

    async def query_user_traffic(self, username: str, reset: bool = False) -> dict[str, int]:
        """Query traffic for a user across all inbounds. Returns {up, down}."""
        up = down = 0
        for suffix in EMAIL_SUFFIXES:
            email = f"{username}@{suffix}"
            for direction in ("uplink", "downlink"):
                pattern = f"user>>>{email}>>>traffic>>>{direction}"
                val = await self._query_single_stat(pattern, reset=reset)
                if direction == "uplink":
                    up += val
                else:
                    down += val
        return {"up": up, "down": down}

    async def query_all_traffic(self) -> dict[str, dict[str, int]]:
        """Batch query all user traffic via single statsquery call."""
        rc, stdout, _ = await self._exec_xray(
            "api", "statsquery", f"--server={self._server}", "-pattern=user>>>",
            timeout=10.0,
        )
        result: dict[str, dict[str, int]] = {}
        if rc != 0 or not stdout:
            return result

        current_name = ""
        for line in stdout.splitlines():
            line = line.strip()
            if line.startswith("name:"):
                current_name = line.split('"')[1] if '"' in line else ""
            elif line.startswith("value:") and current_name:
                val = int(line.split(":", 1)[1].strip())
                parts = current_name.split(">>>")
                if len(parts) == 4:
                    uname = parts[1].split("@")[0]
                    if uname not in result:
                        result[uname] = {"up": 0, "down": 0}
                    key = "up" if parts[3] == "uplink" else "down"
                    result[uname][key] += val
                current_name = ""
        return result

    async def get_sys_stats(self) -> dict:
        """Get xray system stats (goroutines, memory, etc.)."""
        rc, stdout, _ = await self._exec_xray(
            "api", "statssys", f"--server={self._server}",
        )
        if rc != 0 or not stdout:
            return {}
        stats: dict = {}
        for line in stdout.splitlines():
            line = line.strip()
            if ":" in line:
                k, v = line.split(":", 1)
                k, v = k.strip(), v.strip()
                try:
                    stats[k] = int(v)
                except ValueError:
                    stats[k] = v
        return stats

    # ── Health ──

    async def health_check(self) -> bool:
        """Verify xray is running and gRPC API responds."""
        if not await self.is_running():
            return False
        # Quick gRPC probe — statssys always works if API is up
        rc, _, _ = await self._exec_xray(
            "api", "statssys", f"--server={self._server}", timeout=3.0,
        )
        return rc == 0

    async def is_running(self) -> bool:
        """Check if xray container is running."""
        try:
            proc = await asyncio.create_subprocess_exec(
                "docker", "inspect", "-f", "{{.State.Running}}", self.container,
                stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
            )
            stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=5.0)
            return proc.returncode == 0 and b"true" in stdout
        except Exception:
            return False

    async def reload(self) -> bool:
        """Reload xray config via SIGHUP."""
        try:
            proc = await asyncio.create_subprocess_exec(
                "docker", "kill", "--signal=SIGHUP", self.container,
                stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
            )
            await asyncio.wait_for(proc.communicate(), timeout=5.0)
            return proc.returncode == 0
        except Exception as e:
            logger.error("reload failed: %s", e)
            return False

    async def restart(self) -> bool:
        """Restart xray container."""
        try:
            proc = await asyncio.create_subprocess_exec(
                "docker", "restart", self.container,
                stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
            )
            await asyncio.wait_for(proc.communicate(), timeout=30.0)
            return proc.returncode == 0
        except Exception as e:
            logger.error("restart failed: %s", e)
            return False

    # ── Docker validation ──

    async def _check_docker(self) -> bool:
        """Verify Docker CLI is available. Result is cached for the lifetime of this instance."""
        if self._docker_available is not None:
            return self._docker_available
        try:
            proc = await asyncio.create_subprocess_exec(
                "docker", "version",
                stdout=asyncio.subprocess.PIPE,
                stderr=asyncio.subprocess.PIPE,
            )
            await asyncio.wait_for(proc.communicate(), timeout=5.0)
            self._docker_available = proc.returncode == 0
        except FileNotFoundError:
            self._docker_available = False
        except Exception:
            self._docker_available = False

        if not self._docker_available:
            logger.warning("Docker CLI not available — xray API calls will be no-ops")
        return self._docker_available

    # ── Internal ──

    async def _exec_xray(self, *args: str, timeout: float = 5.0) -> tuple[int, str, str]:
        """Run `docker exec <container> xray <args>`, return (rc, stdout, stderr)."""
        if not await self._check_docker():
            return -1, "", "docker not available"
        try:
            proc = await asyncio.create_subprocess_exec(
                "docker", "exec", self.container, "xray", *args,
                stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
            )
            stdout_b, stderr_b = await asyncio.wait_for(proc.communicate(), timeout=timeout)
            return proc.returncode or 0, stdout_b.decode(), stderr_b.decode()
        except asyncio.TimeoutError:
            logger.warning("xray command timed out: %s", " ".join(args[:3]))
            return -1, "", "timeout"
        except FileNotFoundError:
            logger.debug("Docker CLI not available")
            return -1, "", "docker not found"
        except Exception as e:
            logger.debug("xray exec error: %s", e)
            return -1, "", str(e)

    async def _query_single_stat(self, pattern: str, reset: bool = False) -> int:
        """Query a single stat value."""
        cmd = ["api", "statsquery", f"--server={self._server}", f"-pattern={pattern}"]
        if reset:
            cmd.append("-reset")
        rc, stdout, _ = await self._exec_xray(*cmd)
        if rc == 0 and stdout:
            for line in stdout.splitlines():
                line = line.strip()
                if line.startswith("value:"):
                    return int(line.split(":", 1)[1].strip())
        return 0

File: backend/crates/chameleon-vpn/src/xray_api.rs
//! Xray gRPC API client — direct gRPC connection (no Docker CLI dependency).
//!
//! Uses the `xray-core` crate for pre-compiled protobuf bindings.
//! Connects to xray's gRPC API (default: xray:10085 on Docker network).

use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::Mutex;
use tonic::transport::Channel;
use tracing::{debug, error, info, warn};

use xray_core::app::proxyman::command::{
    handler_service_client::HandlerServiceClient,
    AlterInboundRequest,
};
use xray_core::app::stats::command::{
    stats_service_client::StatsServiceClient,
    QueryStatsRequest, SysStatsRequest,
};

/// Inbound tag definitions — email suffix and flow for each inbound.
pub struct InboundDef {
    pub tag: &'static str,
    pub suffix: &'static str,
    pub flow: &'static str,
}

pub static INBOUND_DEFS: &[InboundDef] = &[
    InboundDef { tag: "VLESS TCP REALITY", suffix: "xray", flow: "xtls-rprx-vision" },
    InboundDef { tag: "VLESS XHTTP REALITY", suffix: "xhttp", flow: "" },
    InboundDef { tag: "VLESS XHTTP H2 REALITY", suffix: "xhttp2", flow: "" },
    InboundDef { tag: "VLESS WS CDN", suffix: "ws", flow: "" },
];

/// gRPC-based xray API client.
pub struct XrayApi {
    addr: String,
    stats: Arc<Mutex<Option<StatsServiceClient<Channel>>>>,
    handler: Arc<Mutex<Option<HandlerServiceClient<Channel>>>>,
}

impl XrayApi {
    pub fn new(addr: &str) -> Self {
        // Validate address format
        assert!(
            addr.contains(':') && addr.len() <= 256,
            "Invalid xray gRPC address: {addr}"
        );
        Self {
            addr: addr.to_string(),
            stats: Arc::new(Mutex::new(None)),
            handler: Arc::new(Mutex::new(None)),
        }
    }

    /// Create a lazy channel to xray gRPC. Connection is established on first request.
    fn get_channel(&self) -> Result<Channel, tonic::transport::Error> {
        let endpoint = format!("http://{}", self.addr);
        Ok(Channel::from_shared(endpoint)
            .expect("valid URI")
            .connect_timeout(std::time::Duration::from_secs(5))
            .timeout(std::time::Duration::from_secs(10))
            .connect_lazy())
    }

    async fn stats_client(&self) -> Option<StatsServiceClient<Channel>> {
        let mut guard = self.stats.lock().await;
        if guard.is_none() {
            match self.get_channel() {
                Ok(ch) => *guard = Some(StatsServiceClient::new(ch)),
                Err(e) => { warn!(error = %e, "Failed to create xray stats channel"); return None; }
            }
        }
        guard.clone()
    }

    async fn handler_client(&self) -> Option<HandlerServiceClient<Channel>> {
        let mut guard = self.handler.lock().await;
        if guard.is_none() {
            match self.get_channel() {
                Ok(ch) => *guard = Some(HandlerServiceClient::new(ch)),
                Err(e) => { warn!(error = %e, "Failed to create xray handler channel"); return None; }
            }
        }
        guard.clone()
    }

    /// Reset cached connections (e.g. after xray restart).
    pub async fn reset_connections(&self) {
        *self.stats.lock().await = None;
        *self.handler.lock().await = None;
    }

    // ── User Management ──

    pub async fn add_user(&self, inbound_tag: &str, uuid: &str, email: &str, flow: &str) -> bool {
        let Some(mut client) = self.handler_client().await else { return false; };

        use prost::Message;

        // Build VLESS Account as protobuf
        let account = xray_core::proxy::vless::Account {
            id: uuid.to_string(),
            flow: flow.to_string(),
            encryption: "none".to_string(),
        };

        // Encode account to protobuf bytes
        let mut account_bytes = Vec::new();
        account.encode(&mut account_bytes).unwrap_or_default();

        // Wrap in AddUserOperation
        let op = xray_core::app::proxyman::command::AddUserOperation {
            user: Some(xray_core::common::protocol::User {
                level: 0,
                email: email.to_string(),
                account: Some(xray_core::common::serial::TypedMessage {
                    r#type: "xray.proxy.vless.Account".to_string(),
                    value: account_bytes,
                }),
            }),
        };

        // Encode operation to protobuf bytes
        let mut op_bytes = Vec::new();
        op.encode(&mut op_bytes).unwrap_or_default();

        let request = AlterInboundRequest {
            tag: inbound_tag.to_string(),
            operation: Some(xray_core::common::serial::TypedMessage {
                r#type: "xray.app.proxyman.command.AddUserOperation".to_string(),
                value: op_bytes,
            }),
        };

        match client.alter_inbound(request).await {
            Ok(_) => true,
            Err(e) => {
                warn!(tag = inbound_tag, email, error = %e, "add_user gRPC failed");
                if e.code() == tonic::Code::Unavailable {
                    self.reset_connections().await;
                }
                false
            }
        }
    }

    pub async fn remove_user(&self, inbound_tag: &str, email: &str) -> bool {
        let Some(mut client) = self.handler_client().await else { return false; };

        use prost::Message;

        let op = xray_core::app::proxyman::command::RemoveUserOperation {
            email: email.to_string(),
        };

        let mut op_bytes = Vec::new();
        op.encode(&mut op_bytes).unwrap_or_default();

        let request = AlterInboundRequest {
            tag: inbound_tag.to_string(),
            operation: Some(xray_core::common::serial::TypedMessage {
                r#type: "xray.app.proxyman.command.RemoveUserOperation".to_string(),
                value: op_bytes,
            }),
        };

        match client.alter_inbound(request).await {
            Ok(_) => true,
            Err(e) => {
                debug!(tag = inbound_tag, email, error = %e, "remove_user gRPC failed");
                false
            }
        }
    }

    pub async fn add_user_to_all_inbounds(&self, uuid: &str, username: &str, _short_id: &str) -> bool {
        let mut ok = 0usize;
        for d in INBOUND_DEFS {
            let email = format!("{username}@{}", d.suffix);
            if self.add_user(d.tag, uuid, &email, d.flow).await { ok += 1; }
        }
        if ok < INBOUND_DEFS.len() {
            warn!(username, ok, total = INBOUND_DEFS.len(), "add_user_to_all: partial success");
        }
        ok > 0
    }

    pub async fn remove_user_from_all_inbounds(&self, username: &str) -> bool {
        let mut any_ok = false;
        for d in INBOUND_DEFS {
            let email = format!("{username}@{}", d.suffix);
            if self.remove_user(d.tag, &email).await { any_ok = true; }
        }
        any_ok
    }

    // ── Stats ──

    pub async fn query_all_traffic(&self) -> HashMap<String, TrafficStats> {
        let Some(mut client) = self.stats_client().await else {
            return HashMap::new();
        };

        let request = QueryStatsRequest {
            pattern: "user>>>".to_string(),
            reset: false,
        };

        let response = match client.query_stats(request).await {
            Ok(r) => r.into_inner(),
            Err(e) => {
                warn!(error = %e, "query_all_traffic gRPC failed");
                return HashMap::new();
            }
        };

        let mut result: HashMap<String, TrafficStats> = HashMap::new();
        for stat in response.stat {
            // Format: "user>>>email>>>traffic>>>uplink/downlink"
            let parts: Vec<&str> = stat.name.split(">>>").collect();
            if parts.len() == 4 {
                let uname = parts[1].split('@').next().unwrap_or("");
                let entry = result.entry(uname.to_string()).or_default();
                if parts[3] == "uplink" { entry.up += stat.value; }
                else { entry.down += stat.value; }
            }
        }
        result
    }

    /// Query total server traffic (all users combined).
    /// Returns (total_up, total_down) in bytes.
    pub async fn query_total_traffic(&self) -> (i64, i64) {
        let Some(mut client) = self.stats_client().await else {
            return (0, 0);
        };

        let request = QueryStatsRequest {
            pattern: "user>>>".to_string(),
            reset: false,
        };

        let response = match client.query_stats(request).await {
            Ok(r) => r.into_inner(),
            Err(e) => {
                warn!(error = %e, "query_total_traffic gRPC failed");
                return (0, 0);
            }
        };

        let mut up: i64 = 0;
        let mut down: i64 = 0;
        for stat in response.stat {
            let parts: Vec<&str> = stat.name.split(">>>").collect();
            if parts.len() == 4 {
                if parts[3] == "uplink" { up += stat.value; }
                else { down += stat.value; }
            }
        }
        (up, down)
    }

    /// Count online users — users with any traffic in xray stats.
    /// This counts distinct usernames that have non-zero traffic counters.
    pub async fn count_online_users(&self) -> i32 {
        let traffic = self.query_all_traffic().await;
        traffic.values().filter(|s| s.up > 0 || s.down > 0).count() as i32
    }

    // ── Health ──

    pub async fn health_check(&self) -> bool {
        let Some(mut client) = self.stats_client().await else { return false; };
        match client.get_sys_stats(SysStatsRequest {}).await {
            Ok(_) => true,
            Err(e) => {
                warn!(error = %e, "Xray health check gRPC failed");
                // Reset so next call creates fresh channel
                self.reset_connections().await;
                false
            }
        }
    }

    /// Reset connections so next gRPC call reconnects fresh.
    /// Config file is already written by the caller — this ensures gRPC clients reconnect.
    pub async fn reload(&self) -> bool {
        self.reset_connections().await;
        // Give xray a moment, then verify
        tokio::time::sleep(std::time::Duration::from_millis(500)).await;
        self.health_check().await
    }
}

#[derive(Debug, Default, Clone)]
pub struct TrafficStats {
    pub up: i64,
    pub down: i64,
}
